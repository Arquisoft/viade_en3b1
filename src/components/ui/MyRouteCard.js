import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
import Typography from '@material-ui/core/Typography';
import MapSnapshot from '../map/MapSnapshot.js'
import Button from '@material-ui/core/Button';
import { Link, Grid } from '@material-ui/core';

const useStyles = makeStyles({
  root: {
    maxWidth: 290,
    marginBottom: '2vh',
    marginTop: '2vh',
    background: '#f7f7f7'
  },
  pos: {
    // marginBottom: 12,
    // marginTop: 12,
    color: 'black'
  },
  link: {
    // marginBottom: 20,
    // marginTop: 10,
  },
});

function getDate(){
  let dates = ['2/4/2020', '12/3/2020', '1/4/2020', '6/3/2020', '22/2/2020'];
  let randomElement = dates[Math.floor(Math.random() * dates.length)];
  // console.log("SELECTED DATE: "+randomElement);
  return randomElement;
}

function getDistance(){
  let dist = ['3.63km', '6.3km', '25.4km', '7.96km', '5.42km'];
  let randomElement = dist[Math.floor(Math.random() * dist.length)];
  // console.log("SELECTED DIST: "+randomElement);
  return randomElement;
}

export default function MyRouteCard(props) {
  const classes = useStyles();
  // console.log(props.route);
  return (
    <Card elevation={5} className={classes.root}>
      <CardContent>
        <Grid container spacing={2}>

          <Grid item xs={12}>
            <MapSnapshot route={props.route}></MapSnapshot>
          </Grid>

          <Grid item xs={12} >
            <Link underline='none' href={"#/RouteDetails/" + props.route.getId()}>
              <Typography className={classes.link} variant="h5" component="h2">
                {props.route.getName()}
              </Typography>
            </Link>
          </Grid>

          <Grid item xs={12} sm={5}>
            <Typography className={classes.pos} color="textSecondary">
              {/* {props.route.getDate()} */}
              {getDate()}
            </Typography>
          </Grid>

          <Grid item xs={12} sm={5}>
            <Typography className={classes.pos} color="textSecondary">
              {/* {props.route.getTotalDistance()} */}
              {getDistance()}
            </Typography>
          </Grid>

          <Button variant='outlined'
            // style={{ color: 'white', background: 'black' }} 
            href={"#/RouteDetails/" + props.route.getId()}
            color="inherit"
            style={{marginLeft:'auto', marginRight:'auto'}}>
            Details
          </Button>

        </Grid>
      </CardContent>
    </Card >
  );
}