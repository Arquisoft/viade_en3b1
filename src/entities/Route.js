import { v4 as uuid } from "uuid";
import RouteElement from "./RouteElement";

class Route {

    /**
     * Constructor for a new Route. 
     * An id will be automatically generated (uuid v4).
     * TotalDistance will also be automatically generated.
     * 
     * RouteElements have name, latitude, longitude and elevation.
     * 
     * @param {String} name Name of the route.
     * @param {Date} date Date of the route.
     * @param {String} description Description of the route.
     * @param {Array} routeElements List of RouteElement objects, which are the trackpoints.
     * @param {Array} comments List of comments
     * @param {Array<String>} media List of urls of media (on the POD)
     */
    constructor(name, description, routeElements, comments, media, date) {
        this.id = uuid().toString();
        this.name = name;
        this.date = date;
        this.description = description;

        this.media = media;
        this.totalDistance = null;

        if (comments === null) {
            this.comments = [];
        } else {
            this.comments = comments;
        }

        if (routeElements[0] instanceof RouteElement) {
            this.routeElements = routeElements;
        } else {
            this.routeElements = generateRouteElements(routeElements);
        }
    }

    getId() {
        return this.id;
    }

    getName() {
        return this.name;
    }

    getDate() {
        return this.date;
    }

    getTime() {
        return this.time;
    }

    getDescription() {
        return this.description;
    }

    getRouteElements() {
        return this.routeElements;
    }

    getTotalDistance() {
        var total = 0;
        var aux = 0;

        for (let i = 1; i < this.routeElements.length; i++) {
            aux = this.routeElements[i].getLatitude() - this.routeElements[i - 1].getLatitude();
            total += this.addDistance(aux);
            aux = this.routeElements[i].getLongitude() - this.routeElements[i - 1].getLongitude();
            total += this.addDistance(aux);
            aux = this.routeElements[i].getElevation() - this.routeElements[i - 1].getElevation();
            total += this.addDistance(aux);
        }

        return total;
    }

    addDistance(aux) {
        if (aux < 0) {
            return -aux;
        } else {
            return aux;
        }
    }

    getComments() {
        return this.comments;
    }

    getMedia() {
        return this.media;
    }

    addMedia(media) {
        this.media.push(media);
    }

    addRouteElement(routeElement) {
        this.routeElements.push(routeElement);
    }

    getJsonLD() {
        let routePointsJson = [];
        this.routeElements.forEach((p) => routePointsJson.push(p.toJsonLatLng()));
        return JSON.stringify(
            {
                "@context": {
                    "@version": 1.1,
                    "comments": {
                        "@container": "@list",
                        "@id": "viade:comments"
                    },
                    "description": {
                        "@id": "schema:description",
                        "@type": "xs:string"
                    },
                    "media": {
                        "@container": "@list",
                        "@id": "viade:media"
                    },
                    "name": {
                        "@id": "schema:name",
                        "@type": "xs:string"
                    },
                    "points": {
                        "@container": "@list",
                        "@id": "viade:points"
                    },
                    "latitude": {
                        "@id": "schema:latitude",
                        "@type": "xs:double"
                    },
                    "longitude": {
                        "@id": "schema:longitude",
                        "@type": "xs:double"
                    },
                    "elevation": {
                        "@id": "schema:elevation",
                        "@type": "xsd:double"
                    },
                    "author": {
                        "@id": "schema:author",
                        "@type": "@id"
                    },
                    "rdf": "http://www.w3.org/1999/02/22-rdf-syntax-ns#",
                    "rdfs": "http://www.w3.org/2000/01/rdf-schema#",
                    "schema": "http://schema.org/",
                    "viade": "http://arquisoft.github.io/viadeSpec/",
                    "xsd": "http://www.w3.org/2001/XMLSchema#"
                },
                "name": this.name,
                "author": this.author,
                "description": this.description,
                "comments": this.comments,
                "media": this.media,
                "points": routePointsJson
            }
        );
    }

}

/**
 * Method that receives a list of points and returns
 * a list of RouteElements
 * @param {Array} points List of points containing name, latitude, longitude, elevation. 
 */
function generateRouteElements(points) {
    let routeElements = [];
    points.forEach((p) => routeElements.push(new RouteElement(p.lat, p.lng)));
    return routeElements;
}

export default Route;