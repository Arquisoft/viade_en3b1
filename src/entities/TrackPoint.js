class TrackPoint {

    /**
     * 
     * @param {Number} latitude 
     * @param {Number} longitude 
     * @param {String} name 
     * @param {Number} elevation 
     */
    constructor(latitude, longitude, name, elevation) {

        this.latitude = latitude;
        this.longitude = longitude;

        if (!name) {
            this.name = "";
        } else {
            this.name = name;
        }

        if(!elevation){
            this.elevation = "";
        } else {
            this.elevation = elevation;
        }
    }

    getName() {
        return this.name;
    }

    getLatitude() {
        return this.latitude;
    }

    getElevation() {
        return this.elevation;
    }

    getLongitude() {
        return this.longitude;
    }

    setElevation(num) {
        this.elevation = num;
    }

    toJson() {
        return {
            "latitude": this.latitude,
            "longitude": this.longitude,
            "elevation": this.elevation,
        };
    }
}

export default TrackPoint;