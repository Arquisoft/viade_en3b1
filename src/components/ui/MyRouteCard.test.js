import MyRouteCard from './MyRouteCard';
import Route from '../../entities/Route';
import TrackPoint from '../../entities/TrackPoint';

var route1 = new Route("Route1", "Description 1", [new TrackPoint(1,1), new TrackPoint(2,1)], null, null, null, "ID1");

test('Define MyRouteCard', () => {
  expect(MyRouteCard).toBeDefined();
});

test('Renders correctly', () => {
  const wrapper = render(<MyRouteCard route={route1} />);
  expect(wrapper).toMatchSnapshot();
});