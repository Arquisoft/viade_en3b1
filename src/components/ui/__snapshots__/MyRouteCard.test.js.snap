// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly 1`] = `
<div
  class="MuiPaper-root MuiCard-root makeStyles-root-1 MuiPaper-elevation1 MuiPaper-rounded"
>
  <div
    class="MuiCardContent-root"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
      >
        <div />
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
      >
        <a
          class="MuiTypography-root MuiLink-root MuiLink-underlineNone MuiTypography-colorPrimary"
          href="#/RouteDetails/undefined"
        >
          <h2
            class="MuiTypography-root MuiTypography-h5"
          >
            Route1
          </h2>
        </a>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-5"
      >
        <p
          class="MuiTypography-root makeStyles-pos-2 MuiTypography-body1 MuiTypography-colorTextSecondary"
        >
          12/3/2020
        </p>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-5"
      >
        <p
          class="MuiTypography-root makeStyles-pos-2 MuiTypography-body1 MuiTypography-colorTextSecondary"
        >
          3.63km
        </p>
      </div>
      <a
        aria-disabled="false"
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-colorInherit"
        href="#/RouteDetails/undefined"
        style="margin-left:auto;margin-right:auto"
        tabindex="0"
      >
        <span
          class="MuiButton-label"
        >
          Details
        </span>
      </a>
    </div>
  </div>
</div>
`;
